Resolve get subjects issue

yatharthranjan opened the pull request:
This PR is to resolve the issue #33 and is in accordance with the guidelines specified in #42 

Some major changes are as follows-
-
* A new endpoint has been added for getting information of subjects and projects from the Management Portal.
* A client added for interacting with the Management Portal.
* Added additional model and utility java classes for handling data requests and response from the Management Portal.
* Added `oauth-client-util` library for authentication with the Management Portal.
* Management Portal config loaded from a file called `mp_info.yml`
* Updated the `build-utility.gradle` file to include `mp_info.yml` in integration and end-to-end tests.

The following endpoints are active- 
-
* `/api/mp/subjects`  Get all subjects
* `/api/mp/subjects/{id}` Get subject by id
* `/api/mp/projects/{name}/subjects` Get subjects from a project
* `/api/mp/projects/{name}` Get project by name

The following have been implemented but do not provided intended results because of this issue of MP issue [18.](https://github.com/RADAR-CNS/ManagementPortal/issues/18) Once this issue is resolved, this should work.
* `/api/mp/projects` Get all projects

A separate endpoint (`/mp`) has been created for exposing data retrieved from the MP for now. This can later be merged into (`/api`)
Here {id} of subject in Rest-API = {login} of subject in MP
All of this data is retrieved from the MP and handled and exposed through REST-API.

yatharthranjan committed:
Added config and listener classes for Management Portal

yatharthranjan committed:
Added MP client and MP end points and all MP supporting classes

yatharthranjan committed:
Added info for Management Portal Configuration

yatharthranjan committed:
Cleaned the code and fixed consistency issues. Cleared any pre-configured files

yatharthranjan committed:
Merge branch 'resolve_get_subjects_issue' of https://github.com/RADAR-CNS/RADAR-RestApi into resolve_get_subjects_issue
Merge required because updated Readme on github

afolarin committed:
typo

yatharthranjan committed:
Updated integrationTest

yatharthranjan committed:
Merge branch 'resolve_get_subjects_issue' of https://github.com/RADAR-CNS/RADAR-RestApi into resolve_get_subjects_issue

yatharthranjan committed:
Update mp_info.yml

yatharthranjan committed:
Update build-utility.gradle

yatharthranjan committed:
Update build-utility.gradle

yatharthranjan committed:
Update build-utility.gradle

yatharthranjan committed:
Code cleanup

yatharthranjan committed:
Code Cleanup

yatharthranjan committed:
Code cleanup

yatharthranjan committed:
Code cleanup

yatharthranjan committed:
Code cleanup

yatharthranjan committed:
Use Artifactory for oauth-client-util

yatharthranjan committed:
Updates due to changes in Management Portal api

yatharthranjan committed:
Merge branch 'resolve_get_subjects_issue' of https://github.com/RADAR-CNS/RADAR-RestApi into resolve_get_subjects_issue

dennyverbeeck reviewed (commented):
Hmm, this implementation deserializes the response from MP locally before serializing it again to serve it to the client. Furthermore the list of subjects is only fetched once, so updates to the list of subjects would not be propagated to clients.
Perhaps you can consider using a proxy servlet? E.g. [HTTP-Proxy-Servlet](https://github.com/mitre/HTTP-Proxy-Servlet) can be used to directly forward requests to MP, and directly serve the response to the client. No need to even have a MpClient or the domain objects this way. Probably the following configuration would already achieve this (not tested):
```
<servlet>
    <servlet-name>managementportal</servlet-name>
    <servlet-class>org.mitre.dsmiley.httpproxy.ProxyServlet</servlet-class>
    <init-param>
      <param-name>targetUri</param-name>
      <param-value>http://${mpUrl}/api</param-value>
    </init-param>
    <init-param>
      <param-name>log</param-name>
      <param-value>true</param-value>
    </init-param>
</servlet>
<servlet-mapping>
  <servlet-name>managementportal</servlet-name>
  <url-pattern>/api/mp/*</url-pattern>
</servlet-mapping>
```
Now the only configuration necessary is the `mpUrl` variable. The downside of this approach is that you can, as far as i know, only set this variable as a Java system property, i.e. by passing `-DmpUrl=https://some-url/`. You could set the `mpUrl` to an environment variable as well, thus exposing an environment variable to set the MP url, see [this post](https://stackoverflow.com/questions/1404384/referencing-environment-variables-in-web-xml).

yatharthranjan commented:
@dennyverbeeck I understand what you are saying but the data fetched from MP needs to be formatted in Rest API to make it consistent. For example, the `login` in MP is `subjectId` in Rest API. Also the data fetched from MP is just of the subjects. This information will be used by the REST API to query the MongoDb for more results using the subject data received from MP. This is just to get subject information from MP so that the Rest API does not use the find() query on MongoDb as it is not performant. 
The requirement here is not to just output the data from MP endpoint. We also want to process it and for other purposes. What do you think?

dennyverbeeck commented:
Aha ok, there were always discussions about 'forwarding requests to MP', i did not know there were requirements to translate some of the properties as well. Ok never mind then :smile: Btw, not sure if you know about it but, all the way at the bottom of the Readme for MP there is a [small section](https://github.com/RADAR-CNS/ManagementPortal#client-libraries) on generating a complete client for MP automatically. Sorry if you didn't know about it, i didn't realize it. But it can be useful moving forward!

dennyverbeeck commented on src/main/java/org/radarcns/managementportal/Project.java:
```diff
@@ -0,0 +1,163 @@
+    @JsonProperty("location")
+    private final String location;
+    @JsonProperty("attributes")
+    private final List<Tag> attributes;
```
Might as well be `Map<String, String>`

yatharthranjan replied:
Yes I made the changes but not yet pushed.

yatharthranjan replied:
I think the Project still uses the array [] values for attributes, So this should be fine. Only change was in the subject attributes which now uses object {} .

dennyverbeeck commented on src/main/java/org/radarcns/managementportal/Subject.java:
```diff
@@ -0,0 +1,206 @@
+    @JsonProperty("sources")
+    private final List<Source> sources;
+    @JsonProperty("attributes")
+    private final List<Tag> attributes;
```
Might as well be `Map<String, String>`

yatharthranjan commented:
@dennyverbeeck I saw the client after someone mentioned it in a meeting, but by then I had already implemented this. We can look into it when we need more functionality but for now I think this PR should suffice. Thanks for the feedback. The client seems really helpful 👍

yatharthranjan committed:
Update according to MP data model

yatharthranjan committed:
Bug Fixes

dennyverbeeck committed:
Updates for MP api

Projects api in MP now requires projectName instead of ID. The ID already had type String. But to avoid confusion in the future the variable names were changed to reflect projectName instead of projectId.

yatharthranjan committed:
Merge pull request #46 from RADAR-CNS/mp-api-update

Updates for MP api. LGTM

afolarin committed:
Merge branch 'dev' into resolve_get_subjects_issue

yatharthranjan committed:
Update and optimize MP Client

afolarin commented:
@blootsvoets if you have a spare moment could you please review this one.

blootsvoets commented on src/main/java/org/radarcns/managementportal/MpClient.java:
```diff
@@ -0,0 +1,312 @@
+                allSubjects = Subject.getAllSubjectsFromJson(jsonData);
+                LOGGER.info("Retrieved Subjects from MP.");
+                isSubjectsInitialised = true;
+                response.close();
```
response is closed by the try-with-resources block, no need to do it again

blootsvoets replied:
this is also true for all following `response.close()` calls

yatharthranjan replied:
Okay updating now.

blootsvoets commented on src/main/java/org/radarcns/managementportal/Subject.java:
```diff
@@ -0,0 +1,204 @@
+    @JsonIgnore
+    public static Subject getObject(String response) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```
You could use a single `static final ObjectMapper` and configure it in a `static` block, instead of recreating + rebuilding it in every call.

blootsvoets commented on src/main/java/org/radarcns/managementportal/util/UrlDeseralizer.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class UrlDeseralizer extends JsonDeserializer<URL> {
+
+    private Pattern urlPrefix = Pattern.compile("^(https?://|ftp://).*");
```
can be `static final`

blootsvoets commented on src/main/java/org/radarcns/webapp/ManagementPortalEndPoint.java:
```diff
@@ -0,0 +1,222 @@
+ */
+@Api
+@Path("/" + "mp")
+public class ManagementPortalEndPoint {
```
perhaps it already does, but it might be good to mimic the new API in management portal here, as merged from RADAR-CNS/ManagementPortal#101.

yatharthranjan replied:
Yes I think it already does, the requests for subject are based on login and the requests for project are based on project names.

blootsvoets reviewed (approved):
Looks good to me. Some comments for finetuning.

yatharthranjan committed:
Fine-tuned and query project by names based on comments
